include/crocoddyl/multibody/contacts/contact-6d-loop.hxx

```diff
@@ -0,0 +1,274 @@
+
+///////////////////////////////////////////////////////////////////////////////
+// BSD 3-Clause License
+//
+// Copyright (C) 2019-2024, LAAS-CNRS, University of Edinburgh,
+//                          Heriot-Watt University
+// Copyright note valid unless otherwise stated in individual files.
+// All rights reserved.
+///////////////////////////////////////////////////////////////////////////////
+
+#include <pinocchio/algorithm/frames.hpp>
+#include <pinocchio/algorithm/kinematics-derivatives.hpp>
+
+#include "crocoddyl/core/utils/exception.hpp"
+#include "crocoddyl/multibody/contacts/contact-6d.hpp"
+
+namespace crocoddyl {
+
+template <typename Scalar>
+ContactModel6DLoopTpl<Scalar>::ContactModel6DLoopTpl(
+    boost::shared_ptr<StateMultibody> state, const int joint1_id,
+    const SE3 &joint1_placement, const int joint2_id,
+    const SE3 &joint2_placement, const pinocchio::ReferenceFrame ref,
+    const std::size_t nu, const Vector2s &gains)
+    : Base(state, pinocchio::ReferenceFrame::LOCAL, 6, nu),
+      joint1_id_(joint1_id),
+      joint2_id_(joint2_id),
+      joint1_placement_(joint1_placement),
+      joint2_placement_(joint2_placement),
+      gains_(gains) {
+  if (ref != pinocchio::ReferenceFrame::LOCAL) {
+    std::cerr << "Warning: Only reference frame pinocchio::LOCAL is supported "
+                 "for 6D loop contacts"
+              << std::endl;
+  }
+}
+
+template <typename Scalar>
+ContactModel6DLoopTpl<Scalar>::ContactModel6DLoopTpl(
+    boost::shared_ptr<StateMultibody> state, const int joint1_id,
+    const SE3 &joint1_placement, const int joint2_id,
+    const SE3 &joint2_placement, const pinocchio::ReferenceFrame ref,
+    const Vector2s &gains)
+    : Base(state, pinocchio::ReferenceFrame::LOCAL, 6),
+      joint1_id_(joint1_id),
+      joint2_id_(joint2_id),
+      joint1_placement_(joint1_placement),
+      joint2_placement_(joint2_placement),
+      gains_(gains) {
+  if (ref != pinocchio::ReferenceFrame::LOCAL) {
+    std::cerr << "Warning: Only reference frame pinocchio::LOCAL is supported "
+                 "for 6D loop contacts"
+              << std::endl;
+  }
+}
+
+template <typename Scalar>
+ContactModel6DLoopTpl<Scalar>::~ContactModel6DLoopTpl() {}
+
+template <typename Scalar>
+void ContactModel6DLoopTpl<Scalar>::calc(
+    const boost::shared_ptr<ContactDataAbstract> &data,
+    const Eigen::Ref<const VectorXs> &) {
+  Data *d = static_cast<Data *>(data.get());
+  pinocchio::updateFramePlacements<Scalar>(*state_->get_pinocchio().get(),
+                                           *d->pinocchio);
+  d->j1Xf1.noalias() = joint1_placement_.toActionMatrix();
+  d->j2Xf2.noalias() = joint2_placement_.toActionMatrix();
```

Comment:
Indeed, I changed the way values are stored and computed. For the final note, I suggest that we wait for the discussion in case anything changes on that.

Reply:
Adding a comment for the final note. As discussed, we are using parent joint id and relative placement to identify the contact points here. However, Base::id is a FrameIndex not a jointIndex, so it cannot be used directly. I suggest we wait for a larger refactoring to change this properly.